solutionbox/inception/datalab_solutions/inception/_predictor.py

```diff
@@ -82,57 +59,181 @@ def predict(model_dir, image_files):
   
-  results_batches = _tf_predict(model_dir, images)
-  results = _expand_batches(results_batches)
-  input_csv_f = ml.util._file.read_file_stream(input_csv)
-  reader = csv.reader(input_csv_f)
-  include_target = False
-  with ml.util._file.open_local_or_gcs(output_file, mode='w') as f_out:
-    writer = csv.writer(f_out)
-    for input, prediction in zip(reader, results):
-      predicted = prediction[0]
-      labels = list(prediction[1])
-      class_scores = prediction[2]
-      if include_target or len(input) > 1:
-        include_target = True
-        target_index = labels.index(input[1])
-        target_prob = class_scores[target_index]
-      predicted_index = labels.index(predicted)
-      predicted_prob = class_scores[predicted_index]
-      target_prob_value = [str(target_prob)] if include_target else []
-      writer.writerow(input + [predicted] + target_prob_value + [str(predicted_prob)] +
-                      map(str, class_scores))
-  if include_target:
-    schema = [
+  def process(self, context):
+    import google.cloud.ml as ml
+    element = context.element  
+    with ml.util._file.open_local_or_gcs(element['image_url'], 'r') as ff:
+      image_bytes = ff.read()
+    out_element = {'image_bytes': image_bytes}
+    out_element.update(element)
+    yield out_element
+
+
+class PredictBatchDoFn(beam.DoFn):
+  """A DoFn that does batch prediction."""
+
+  def __init__(self, model_dir):
+    import os
+
+    self._model_dir = os.path.join(model_dir, 'model')
+    self._session = None
+    self._tf_inputs = None
+    self._tf_outputs = None
+
+  def start_bundle(self, context):
+    import json
+    import os
+    import tensorflow as tf
+
+    self._session = tf.Session()
+    new_saver = tf.train.import_meta_graph(os.path.join(self._model_dir, 'export.meta'))
+    new_saver.restore(self._session, os.path.join(self._model_dir, 'export'))
+    init_op = tf.get_collection(tf.contrib.session_bundle.constants.INIT_OP_KEY)[0]
+    self._session.run(init_op)
+    self._tf_inputs = json.loads(tf.get_collection('inputs')[0])
+    self._tf_outputs = json.loads(tf.get_collection('outputs')[0])
+
+  def finish_bundle(self, context):
+    if self._session is not None:
+      self._session.close()
+
+  def process(self, context):
+    import collections
+    
+    element = context.element
+    image_urls = [x['image_url'] for x in element]
+    targets = None
+    if 'label' in element[0] and element[0]['label'] is not None:
+      targets = [x['label'] for x in element]
+
+    feed_dict = collections.defaultdict(list)
+    feed_dict[self._tf_inputs['image_bytes']] = [x['image_bytes'] for x in element]
+    feed_dict[self._tf_inputs['key']] = image_urls
+    predictions, labels, scores = self._session.run(
+        [self._tf_outputs['prediction'], self._tf_outputs['labels'], self._tf_outputs['scores']],
+        feed_dict=feed_dict)
+    if targets is not None:
+      yield zip(image_urls, targets, predictions, labels, scores)
```

Comment:
Do these zip's work as expected if feed_dict only has 1 element? Please test as I ran into that issue.